File: web/src/lib/productService.ts
import { db } from './db';
import type { Product } from '@/types';
import { productStockStatus } from '@/types';

const DEFAULT_LOW_STOCK = 5;

function normalizeProduct(p: Product): Product {
  return {
    ...p,
    track_inventory: p.track_inventory ?? false,
    stock_quantity: p.stock_quantity ?? 0,
    low_stock_threshold: p.low_stock_threshold ?? DEFAULT_LOW_STOCK,
  };
}

export async function getAllProducts(activeOnly = false): Promise<Product[]> {
  let list = (await db.products.orderBy('name').toArray()).map(normalizeProduct);
  if (activeOnly) list = list.filter((p) => p.active);
  return list;
}

export async function getProduct(id: number): Promise<Product | undefined> {
  const p = await db.products.get(id);
  return p ? normalizeProduct(p) : undefined;
}

export async function addProduct(input: {
  name: string;
  cost_price: number;
  default_sell_price: number;
  default_commission: number;
  note?: string;
  track_inventory?: boolean;
  stock_quantity?: number;
  low_stock_threshold?: number;
}): Promise<Product> {
  const product: Product = {
    name: input.name.trim(),
    cost_price: input.cost_price,
    default_sell_price: input.default_sell_price,
    default_commission: input.default_commission,
    note: input.note?.trim() || null,
    active: true,
    track_inventory: input.track_inventory ?? false,
    stock_quantity: input.track_inventory ? (input.stock_quantity ?? 0) : 0,
    low_stock_threshold: input.low_stock_threshold ?? DEFAULT_LOW_STOCK,
    created_at: Date.now(),
  };
  const id = await db.products.add(product);
  return { ...product, id };
}

export async function updateProduct(product: Product): Promise<void> {
  if (!product.id) throw new Error('Product id required');
  const normalized = normalizeProduct(product);
  if (!normalized.track_inventory) {
    normalized.stock_quantity = 0;
  }
  await db.products.put(normalized);
}

export async function deleteProduct(id: number): Promise<void> {
  await db.products.delete(id);
}

export async function toggleProductActive(product: Product): Promise<void> {
  const nextActive = !product.active;
  await updateProduct({ ...normalizeProduct(product), active: nextActive });

  if (!nextActive && product.id) {
    const linked = await db.customers.filter((c) => c.product_id === product.id).toArray();
    for (const c of linked) {
      if (c.id) {
        await db.customers.update(c.id, { product_id: null });
      }
    }
  }
}

/** SP đang theo dõi kho và sắp hết / hết hàng */
export async function getLowStockProducts(): Promise<Product[]> {
  const list = await getAllProducts(true);
  return list.filter((p) => {
    const s = productStockStatus(p);
    return s === 'low' || s === 'out';
  });
}

/** Trừ tồn khi ghi đơn (gọi trong transaction) */
export async function deductStock(productId: number, quantity: number): Promise<void> {
  const product = await db.products.get(productId);
  if (!product) return;
  const p = normalizeProduct(product);
  if (!p.track_inventory) return;
  await db.products.put({
    ...p,
    stock_quantity: p.stock_quantity - quantity,
  });
}

/** Hoàn tồn khi huỷ đơn (gọi trong transaction) */
export async function restoreStock(productId: number, quantity: number): Promise<void> {
  const product = await db.products.get(productId);
  if (!product) return;
  const p = normalizeProduct(product);
  if (!p.track_inventory) return;
  await db.products.put({
    ...p,
    stock_quantity: p.stock_quantity + quantity,
  });
}

export function applyProductDefaults(product: Product, quantity = 1) {
  const p = normalizeProduct(product);
  return {
    product_id: p.id,
    product_name: p.name,
    quantity,
    unit_sell_price: p.default_sell_price,
    unit_cost: p.cost_price,
    unit_commission: p.default_commission,
  };
}

export function stockStatusLabel(status: ReturnType<typeof productStockStatus>): string {
  switch (status) {
    case 'out':
      return 'Hết hàng';
    case 'low':
      return 'Sắp hết';
    case 'ok':
      return 'Còn hàng';
    default:
      return 'Không theo kho';
  }
}

File: app/lib/services/product_service.dart
import '../database/database_helper.dart';
import '../models/product.dart';

class ProductService {
  final DatabaseHelper _db = DatabaseHelper();

  Future<List<Product>> getAllProducts({bool activeOnly = false}) =>
      _db.getAllProducts(activeOnly: activeOnly);

  Future<Product?> getProduct(int id) => _db.getProduct(id);

  Future<Product> addProduct({
    required String name,
    required double costPrice,
    required double defaultSellPrice,
    required double defaultCommission,
    String? note,
    bool trackInventory = false,
    int stockQuantity = 0,
    int lowStockThreshold = 5,
  }) async {
    final product = Product(
      name: name.trim(),
      costPrice: costPrice,
      defaultSellPrice: defaultSellPrice,
      defaultCommission: defaultCommission,
      note: note?.trim().isEmpty == true ? null : note?.trim(),
      trackInventory: trackInventory,
      stockQuantity: trackInventory ? stockQuantity : 0,
      lowStockThreshold: lowStockThreshold,
      createdAt: DateTime.now().millisecondsSinceEpoch,
    );
    final id = await _db.insertProduct(product);
    return product.copyWith(id: id);
  }

  Future<void> updateProduct(Product product) async {
    var p = product;
    if (!p.trackInventory) p = p.copyWith(stockQuantity: 0);
    await _db.updateProduct(p);
  }

  Future<void> deleteProduct(int id) => _db.deleteProduct(id);

  Future<void> toggleActive(Product product) async {
    final nextActive = !product.active;
    await updateProduct(product.copyWith(active: nextActive));
    if (!nextActive && product.id != null) {
      final db = await _db.database;
      await db.update(
        'customers',
        {'product_id': null},
        where: 'product_id = ?',
        whereArgs: [product.id],
      );
    }
  }

  Future<List<Product>> getLowStockProducts() async {
    final list = await getAllProducts(activeOnly: true);
    return list
        .where((p) => p.stockStatus == 'low' || p.stockStatus == 'out')
        .toList();
  }

  Future<void> deductStock(int productId, int quantity) async {
    final product = await _db.getProduct(productId);
    if (product == null || !product.trackInventory) return;
    await _db.updateProduct(
      product.copyWith(stockQuantity: product.stockQuantity - quantity),
    );
  }

  Future<void> restoreStock(int productId, int quantity) async {
    final product = await _db.getProduct(productId);
    if (product == null || !product.trackInventory) return;
    await _db.updateProduct(
      product.copyWith(stockQuantity: product.stockQuantity + quantity),
    );
  }

  Map<String, dynamic> applyDefaults(Product product, {int quantity = 1}) {
    return {
      'product_id': product.id,
      'product_name': product.name,
      'quantity': quantity,
      'unit_sell_price': product.defaultSellPrice,
      'unit_cost': product.costPrice,
      'unit_commission': product.defaultCommission,
    };
  }
}
